server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -4181,10 +4183,25 @@ public void waitForSegmentGeneration(long segmentGeneration, ActionListener<Long
         waitForPrimaryTermAndGeneration(getOperationPrimaryTerm(), segmentGeneration, listener);
     }
 
+    private final SubscribableListener<Void> waitForPrimaryTermAndGenerationListeners = new SubscribableListener<Void>();
+
+    private void checkAndCallWaitForPrimaryTermAndGenerationListeners() {
+        if (state == IndexShardState.STARTED || state == IndexShardState.CLOSED) {
+            waitForPrimaryTermAndGenerationListeners.onResponse(null);
+        }
+    }
+
     /**
      * Registers a listener for an event when the shard advances to the provided primary term and segment generation
      */
     public void waitForPrimaryTermAndGeneration(long primaryTerm, long segmentGeneration, ActionListener<Long> listener) {
-        getEngine().addPrimaryTermAndGenerationListener(primaryTerm, segmentGeneration, listener);
+        waitForPrimaryTermAndGenerationListeners.addListener(
+            ActionListener.wrap(
+                (ignored) -> getEngine().addPrimaryTermAndGenerationListener(primaryTerm, segmentGeneration, listener),
+                listener::onFailure
+            )
+        );
+        checkAndCallWaitForPrimaryTermAndGenerationListeners();
```

Comment:
I'm not sure I understand. If the state changes to started/closed, then we should have called `checkAndCallWaitForStartedOrClosedShardListeners` (which is done in close and udpateShardState). And if waitForStartedOrClosedShardListeners is resolved, then adding new listeners to it would immediately resolve the added listener.

Also I'm now wondering, do we need to reset this SubscribableListener? What if we are moving from close to open again?

Reply:
> And if waitForStartedOrClosedShardListeners is resolved, then adding new listeners to it would immediately resolve the added listener.

Ah yes @pxsalehi I missed that point. I can remove the check then here indeed. Done, thanks!

> What if we are moving from close to open again?

There's no going backwards in IndexShard states. A new IndexShard is re-allocated if necessary, which starts again from the CREATED state, goes through recovery, and finally STARTED and ultimately CLOSED.